### ISSUE TITLE
Newly added column is not shown in Azure Storage Explorer for an Azure Table

### ISSUE THREAD

**Issue by shannonlharris** [microsoft/AzureStorageExplorer#356]
I added a new column to an Azure Table Storage. However the new column is not visible in the results grid nor is it available in the Column Options dialog. Refresh doesn't help, Detaching/Re-Attaching doesn't help, Closing/Reopening Storage Explorer doesn't help.

**Storage Explorer Version**: 1.2.0
**Platform**: Windows 10

#### Repro Steps ####

1. Add new Property to Table
2. Verify Property Exists
3. Query (outside of Storage Explorer) to confirm property exists with data

#### Expected Result ####

Any Storage Explorer query for records that include the property should cause the column to be available in the Results grid, column chooser, etc.

#### Actual Result ####

- Attached to Table in Storage Explorer
- New Property/Column never displays in Results Grid nor any Column chooser (Advanced, etc.)

**Comment by craxal** [comment]
Please provide the following information as per our issue template:
- Storage Explorer version
- Platform/OS and Version
- Regression From (last version of Storage Explorer this was working for you)
- Repro Steps (be as specific as possible)
- Expected Result
- Actual Result

**Comment by shannonlharris** [comment]
- Storage Explorer version
      - 1.2.0
   - Platform/OS and Version
      - Windows 10
   - Regression From (last version of Storage Explorer this was working for
   you)
      - unknown/never attempted prior
   - Repro Steps (be as specific as possible)
      - Added new Property to Table
      - Verified Property Exists
      - Queried (outside of Storage Explorer) to confirm property exists
      with data
      -
      - Expected Result
      - Any Storage Explorer query for records that include the property
      should cause the column to be available in the Results grid, column
      chooser, etc.
   - Actual Result
   -
      - Attached to Table in Storage Explorer
      - New Property/Column never displays in Results Grid nor any Column
      chooser (Advanced, etc.)

**Comment by craxal** [comment]
What happens if you refresh the table? How about if you close the editor then reopen?

**Comment by shannonlharris** [comment]
I tried all of the above, but no luck.  Oddly enough, I can copy the query
from the text editor and give it to another user (using the same version)
and they can see the column (though not initially, only after closing and
reopening).

**Comment by craxal** [comment]
How exactly are you adding this new property? Storage Explorer works by listing entities from the table and aggregating the properties from each entity. It's highly likely that, if an entity with a unique property isn't within the first 1000 entities, it's property column won't appear right away. Azure tables aren't like SQL tables, which have discrete schemas.

**Comment by shannonlharris** [comment]
yeah, we definitely have legacy data that doesn't have the property, so
when the table is first accessed in Storage Explorer and it retrieves data
automatically, it would not be a property in those first 1000 entities.
However, my assumption would be that once I query for a specific range of
entities that does have the Property, Storage Explorer would adapt
dynamically since the property is available in the query result set and
update it's result grid.

**Comment by craxal** [comment]
Are you selecting a subset of columns? Is the query using the new property at all? Does the query still return the expected set of entities, even if the column isn't shown?

One way you can tell if you're getting the right data from the server is:
1. Hit F12 to open the dev tools.
2. Open the Sources tab. Hit Ctrl+P, type TableEntityListViewModel.ts, and press Enter.
3. Find the line `result = _a.sent();` (somewhere around line 371) and drop a breakpoint.
4. Rerun your query that includes the entity with the expected property (might want to match partition and row keys for simplicity). The breakpoint should pause execution.
5. Examine the value of `result`. The target entity should have the new property you're expecting.

If this goes as expected, it's likely a bug on our end and we'll need to look into it further. Is this issue blocking you in any way?

**Comment by shannonlharris** [comment]
Ok,

so I've confirmed that the issue is due to how Storage Explorer locks in
the columns based on the first 1000 entities.  Understood that Storage
Explorer needs some mechanism to determine what properties are available
given the schema-less nature of Azure Tables. However, since Storage
Explorer auto-queries and returns the oldest records first when a Table is
opened and before I can specify any custom query, my column choices are
limited to whatever properties are available on the oldest 1000 records.
Any of my subsequent queries in Storage Explorer are limited to those
original properties as columns, as those properties get baked in as the
only possible properties to select.  It would be nice if after opening a
Table, and then upon a subsequent custom query the "Available" Columns for
Selection are updated.

I confirmed this behavior by going back and updating the first 1000 records
in the Table to have the new property, now the new property is available as
a column choice in Storage Explorer.

**Comment by craxal** [comment]
Table editors should already by checking for new columns when it retrieves new items. Is the only issue here that the query builder isn't detecting the new columns?

**Comment by shannonlharris** [comment]
Yes, no editing is being attempted here.

**Comment by craxal** [comment]
Let me clarify. If you leave the query builder alone and add a new entity with a new property, does the new column appear? If not, does loading it into the cache (clicking Load More until you find it) work? Does refreshing work?

Now do the same for when the query builder is open and run. Same questions; what happens?

**Comment by shannonlharris** [comment]
I'll need to get setup in another DEV environment.  To date, we've never
used StorageExplorer for editing, only viewing. At any rate, it's not a
scenario I could repeat on Production data.

**Comment by MRayermannMSFT** [comment]
@shannonlharris, please open a new issue if you're still having problems. Thanks.
